packages/transaction/src/zk-tx.ts

```diff
@@ -363,4 +370,75 @@ export class ZkTx {
       protocol: 'groth16',
     }
   }
+
+  parseMemo(): {
+    encryptedNotes: Buffer[]
+    prepayInfo?: {
+      prepayFeeInEth: Fp
+      prepayFeeInToken: Fp
+      expiration: number
+      signature: Buffer
+    }
+  } {
+    if (!this.memo)
+      return {
+        encryptedNotes: [],
+      }
+    if (this.memo.version === MemoVersion.V1) {
+      return {
+        encryptedNotes: [this.memo.data],
+      }
+    }
+    if (this.memo.version !== MemoVersion.V2) {
+      throw new Error(`Unrecognized memo version: ${this.memo.version}`)
+    }
+    const memoSig = Bytes4.from(
+      `0x${this.memo.data.slice(0, 4).toString('hex')}`,
+    )
+    if (
+      V2_MEMO_DEFAULT_ABI.eq(memoSig) ||
+      V2_MEMO_DEFAULT_ABI_ZERO.eq(memoSig)
+    ) {
+      return {
+        encryptedNotes: Array((this.memo.data.length - 4) / 81)
+          .fill(null)
+          .map((_, index) => {
+            if (!this.memo) throw new Error('Expected memo to exist')
```

Comment:
Is this duplicated short-circuit line 383? 
https://github.com/zkopru-network/zkopru/blob/d6b4f279625c33368946d2979d8fc2ae5cb7ed8f/packages/transaction/src/zk-tx.ts#L383-L386

Reply:
Inside the `map` function typescript will throw an error `Object is possibly undefined`. I think it's a limitation of Typescript, the `map` function could be executed after some other function changes `this.memo`.